[CI] PackageUpgradeTests test12SetupBwcVersion failure

### CI Link

https://gradle-enterprise.elastic.co/s/lb6ayhb4eeqdk

### Repro line

none

### Does it reproduce?

Didn't try

### Applicable branches

7.17

### Failure history

_No response_

### Failure excerpt

This one is common to a bunch of BWC failures, all from 7.17 with BWC_VERSION in 6.x range. And all with `Rocky Linux 8.7, Linux 4.18.0-425.3.1.el8.x86_64`.

https://gradle-enterprise.elastic.co/s/zj4ezikoajsye
https://gradle-enterprise.elastic.co/s/wxb6bjfd2yw4i
https://gradle-enterprise.elastic.co/s/kqi3qj6peu6ma


There is also another freshly created issue which seems to point in the same direction: https://github.com/elastic/elasticsearch/issues/91604

The failure:

```
:qa:os:destructiveDistroUpgradeTest.v6.8.22.default-deb-aarch64#deps
:qa:os:destructiveDistroUpgradeTest.v6.8.22.default-deb-aarch64 SKIPPED
:qa:os:destructiveDistroUpgradeTest.v6.8.22.default-rpm#deps
:qa:os:destructiveDistroUpgradeTest.v6.8.22.default-rpm FAILED

org.elasticsearch.packaging.test.PackageUpgradeTests > test12SetupBwcVersion FAILED
    java.lang.RuntimeException: Elasticsearch did not start
        at __randomizedtesting.SeedInfo.seed([969B78FBC4130EFA:A70A9CED4B442FEA]:0)
        at org.elasticsearch.packaging.util.ServerUtils.waitForElasticsearch(ServerUtils.java:213)
        at org.elasticsearch.packaging.util.ServerUtils.waitForElasticsearch(ServerUtils.java:85)
        at org.elasticsearch.packaging.util.Packages.assertElasticsearchStarted(Packages.java:272)
        at org.elasticsearch.packaging.test.PackagingTestCase.awaitElasticsearchStartup(PackagingTestCase.java:347)
        at org.elasticsearch.packaging.test.PackagingTestCase.startElasticsearch(PackagingTestCase.java:365)
        at org.elasticsearch.packaging.test.PackageUpgradeTests.test12SetupBwcVersion(PackageUpgradeTests.java:53)

        Caused by:
        org.apache.http.conn.HttpHostConnectException: Connect to localhost:9200 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
            at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:156)
            at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:374)
            at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:393)
            at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
            at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
            at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
            at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
            at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
            at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
            at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
            at org.apache.http.client.fluent.Request.internalExecute(Request.java:173)
            at org.apache.http.client.fluent.Executor.execute(Executor.java:262)
            at org.elasticsearch.packaging.util.ServerUtils.execute(ServerUtils.java:133)
            at org.elasticsearch.packaging.util.ServerUtils.waitForElasticsearch(ServerUtils.java:177)
            ... 5 more

            Caused by:
            java.net.ConnectException: Connection refused (Connection refused)
                at java.net.PlainSocketImpl.socketConnect(Native Method)
                at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:476)
                at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:218)
                at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:200)
                at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:394)
                at java.net.Socket.connect(Socket.java:606)
                at org.apache.http.conn.socket.PlainConnectionSocketFactory.connectSocket(PlainConnectionSocketFactory.java:75)
                at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:142)
                ... 18 more
```

Looking deeper at logs:

```
  1> [2022-11-16T10:48:29,042][ERROR][o.e.b.Bootstrap          ] [uNpQXsl] Exception
  1> java.lang.IllegalStateException: failed to load plugin class [org.elasticsearch.xpack.core.XPackPlugin]
  1> 	at org.elasticsearch.plugins.PluginsService.loadPlugin(PluginsService.java:616) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.plugins.PluginsService.loadBundle(PluginsService.java:558) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.plugins.PluginsService.loadBundles(PluginsService.java:473) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.plugins.PluginsService.<init>(PluginsService.java:163) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.node.Node.<init>(Node.java:339) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.node.Node.<init>(Node.java:266) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:212) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:212) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:333) [elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:159) [elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:150) [elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86) [elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124) [elasticsearch-cli-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.cli.Command.main(Command.java:90) [elasticsearch-cli-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:116) [elasticsearch-6.8.22.jar:6.8.22]
  1> 	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:93) [elasticsearch-6.8.22.jar:6.8.22]
  1> Caused by: java.lang.reflect.InvocationTargetException
  1> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
  1> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
  1> 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
  1> 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_352]
  1> 	at org.elasticsearch.plugins.PluginsService.loadPlugin(PluginsService.java:607) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	... 15 more
  1> Caused by: java.lang.ExceptionInInitializerError
  1> 	at sun.security.ssl.TrustStoreManager.getTrustedCerts(TrustStoreManager.java:56) ~[?:?]
  1> 	at sun.security.ssl.TrustManagerFactoryImpl.engineInit(TrustManagerFactoryImpl.java:49) ~[?:?]
  1> 	at javax.net.ssl.TrustManagerFactory.init(TrustManagerFactory.java:250) ~[?:1.8.0_352]
  1> 	at org.elasticsearch.xpack.core.ssl.CertParsingUtils.trustManager(CertParsingUtils.java:279) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.ssl.DefaultJDKTrustConfig.createTrustManager(DefaultJDKTrustConfig.java:48) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.ssl.SSLService.createSslContext(SSLService.java:394) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.ssl.SSLService.loadSSLConfigurations(SSLService.java:427) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.ssl.SSLService.<init>(SSLService.java:103) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.XPackPlugin.<init>(XPackPlugin.java:144) ~[?:?]
  1> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
  1> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
  1> 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
  1> 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_352]
  1> 	at org.elasticsearch.plugins.PluginsService.loadPlugin(PluginsService.java:607) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	... 15 more
  1> Caused by: java.security.AccessControlException: access denied ("java.io.FilePermission" "/etc/pki/java/cacerts" "read")
  1> 	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472) ~[?:1.8.0_352]
  1> 	at java.security.AccessController.checkPermission(AccessController.java:886) ~[?:1.8.0_352]
  1> 	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549) ~[?:1.8.0_352]
  1> 	at java.lang.SecurityManager.checkRead(SecurityManager.java:888) ~[?:1.8.0_352]
  1> 	at java.io.File.exists(File.java:825) ~[?:1.8.0_352]
  1> 	at sun.security.tools.KeyStoreUtil.getCacertsKeyStoreFile(KeyStoreUtil.java:141) ~[?:?]
  1> 	at sun.security.ssl.TrustStoreManager$TrustStoreDescriptor.<clinit>(TrustStoreManager.java:80) ~[?:?]
  1> 	at sun.security.ssl.TrustStoreManager.getTrustedCerts(TrustStoreManager.java:56) ~[?:?]
  1> 	at sun.security.ssl.TrustManagerFactoryImpl.engineInit(TrustManagerFactoryImpl.java:49) ~[?:?]
  1> 	at javax.net.ssl.TrustManagerFactory.init(TrustManagerFactory.java:250) ~[?:1.8.0_352]
  1> 	at org.elasticsearch.xpack.core.ssl.CertParsingUtils.trustManager(CertParsingUtils.java:279) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.ssl.DefaultJDKTrustConfig.createTrustManager(DefaultJDKTrustConfig.java:48) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.ssl.SSLService.createSslContext(SSLService.java:394) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.ssl.SSLService.loadSSLConfigurations(SSLService.java:427) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.ssl.SSLService.<init>(SSLService.java:103) ~[?:?]
  1> 	at org.elasticsearch.xpack.core.XPackPlugin.<init>(XPackPlugin.java:144) ~[?:?]
  1> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
  1> 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
  1> 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
  1> 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_352]
  1> 	at org.elasticsearch.plugins.PluginsService.loadPlugin(PluginsService.java:607) ~[elasticsearch-6.8.22.jar:6.8.22]
  1> 	... 15 more
```

--- a/.idea/checkstyle-idea.xml
+++ b/.idea/checkstyle-idea.xml
@@ -1,20 +1,20 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-    <component name="CheckStyle-IDEA" serialisationVersion="2">
-        <checkstyleVersion>10.3.1</checkstyleVersion>
-        <scanScope>JavaOnlyWithTests</scanScope>
-        <option name="thirdPartyClasspath">
-            <option value="$PROJECT_DIR$/build-conventions/build/libs/build-conventions.jar" />
-        </option>
-        <option name="activeLocationIds">
-            <option value="dfb90780-ced3-4152-a4df-3cc9be08e3d8" />
-        </option>
-        <option name="locations">
-            <list>
-                <ConfigurationLocation id="bundled-sun-checks" type="BUNDLED" scope="All" description="Sun Checks">(bundled)</ConfigurationLocation>
-                <ConfigurationLocation id="bundled-google-checks" type="BUNDLED" scope="All" description="Google Checks">(bundled)</ConfigurationLocation>
-                <ConfigurationLocation id="dfb90780-ced3-4152-a4df-3cc9be08e3d8" type="PROJECT_RELATIVE" scope="All" description="Elasticsearch">$PROJECT_DIR$/checkstyle_ide.xml</ConfigurationLocation>
-            </list>
-        </option>
-    </component>
+  <component name="CheckStyle-IDEA" serialisationVersion="2">
+    <checkstyleVersion>10.3.1</checkstyleVersion>
+    <scanScope>JavaOnlyWithTests</scanScope>
+    <option name="thirdPartyClasspath">
+      <option value="$PROJECT_DIR$/build-conventions/build/libs/build-conventions.jar" />
+    </option>
+    <option name="activeLocationIds">
+      <option value="dfb90780-ced3-4152-a4df-3cc9be08e3d8" />
+    </option>
+    <option name="locations">
+      <list>
+        <ConfigurationLocation id="bundled-sun-checks" type="BUNDLED" scope="All" description="Sun Checks">(bundled)</ConfigurationLocation>
+        <ConfigurationLocation id="bundled-google-checks" type="BUNDLED" scope="All" description="Google Checks">(bundled)</ConfigurationLocation>
+        <ConfigurationLocation id="dfb90780-ced3-4152-a4df-3cc9be08e3d8" type="PROJECT_RELATIVE" scope="All" description="Elasticsearch">$PROJECT_DIR$/checkstyle_ide.xml</ConfigurationLocation>
+      </list>
+    </option>
+  </component>
 </project>

--- a/qa/os/src/test/java/org/elasticsearch/packaging/util/Distribution.java
+++ b/qa/os/src/test/java/org/elasticsearch/packaging/util/Distribution.java
@@ -47,7 +47,7 @@ public class Distribution {
             tmpVersion = tmpVersion.replace(".rpm", "");
         }
         this.baseVersion = tmpVersion;
-        this.hasJdk = filename.contains("no-jdk") == false && Version.fromString(baseVersion).onOrAfter(Version.V_7_0_0);
+        this.hasJdk = isDocker() || (filename.contains("no-jdk") == false && Version.fromString(baseVersion).onOrAfter(Version.V_7_0_0));
         this.version = filename.contains("-SNAPSHOT") ? this.baseVersion + "-SNAPSHOT" : this.baseVersion;
     }
 

--- a/qa/os/src/test/java/org/elasticsearch/packaging/util/Distribution.java
+++ b/qa/os/src/test/java/org/elasticsearch/packaging/util/Distribution.java
@@ -8,6 +8,8 @@
 
 package org.elasticsearch.packaging.util;
 
+import org.elasticsearch.Version;
+
 import java.nio.file.Path;
 import java.util.Locale;
 
@@ -38,7 +40,6 @@ public class Distribution {
         }
 
         this.platform = filename.contains("windows") ? Platform.WINDOWS : Platform.LINUX;
-        this.hasJdk = filename.contains("no-jdk") == false;
         String tmpVersion = filename.split("-", 3)[1];
         if (packaging == Packaging.DEB) {
             tmpVersion = tmpVersion.replace(".deb", "");
@@ -46,6 +47,7 @@ public class Distribution {
             tmpVersion = tmpVersion.replace(".rpm", "");
         }
         this.baseVersion = tmpVersion;
+        this.hasJdk = filename.contains("no-jdk") == false && Version.fromString(baseVersion).onOrAfter(Version.V_7_0_0);
         this.version = filename.contains("-SNAPSHOT") ? this.baseVersion + "-SNAPSHOT" : this.baseVersion;
     }
 

--- a/qa/os/src/test/java/org/elasticsearch/packaging/util/Packages.java
+++ b/qa/os/src/test/java/org/elasticsearch/packaging/util/Packages.java
@@ -11,6 +11,7 @@ package org.elasticsearch.packaging.util;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.Version;
+import org.elasticsearch.common.recycler.Recycler;
 import org.elasticsearch.packaging.util.Shell.Result;
 
 import java.io.IOException;
@@ -79,8 +80,9 @@ public class Packages {
 
     public static Installation installPackage(Shell sh, Distribution distribution) throws IOException {
         String systemJavaHome = sh.run("echo $SYSTEM_JAVA_HOME").stdout.trim();
+        String javaHomeEnvVar = Version.fromString(distribution.baseVersion).onOrAfter(Version.V_7_12_0) ? "ES_JAVA_HOME" : "JAVA_HOME";
         if (distribution.hasJdk == false) {
-            sh.getEnv().put("ES_JAVA_HOME", systemJavaHome);
+            sh.getEnv().put(javaHomeEnvVar, systemJavaHome);
         }
         final Result result = runPackageManager(distribution, sh, PackageManagerCommand.INSTALL);
         if (result.exitCode != 0) {
@@ -90,7 +92,7 @@ public class Packages {
         Installation installation = Installation.ofPackage(sh, distribution);
 
         if (distribution.hasJdk == false) {
-            Files.write(installation.envFile, singletonList("ES_JAVA_HOME=" + systemJavaHome), StandardOpenOption.APPEND);
+            Files.write(installation.envFile, singletonList(javaHomeEnvVar + "=" + systemJavaHome), StandardOpenOption.APPEND);
         }
 
         if (Version.fromString(distribution.baseVersion).onOrAfter(Version.V_7_13_0)) {

--- a/qa/os/src/test/java/org/elasticsearch/packaging/util/Packages.java
+++ b/qa/os/src/test/java/org/elasticsearch/packaging/util/Packages.java
@@ -11,7 +11,6 @@ package org.elasticsearch.packaging.util;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.Version;
-import org.elasticsearch.common.recycler.Recycler;
 import org.elasticsearch.packaging.util.Shell.Result;
 
 import java.io.IOException;

